### PR TITLE
RUST-217 Rephrase unsupported/corrupt BSON messages

### PR DESCRIPTION
**Pull request by patrickfreed**
[RUST-217](https://jira.mongodb.org/browse/RUST-217)

This PR adds the key to the error returned when encountering an unsupported/corrupt BSON type.

### CODE REVIEW

File: src/decoder/error.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ use serde::de::{self, Expected, Unexpected};
 pub enum DecoderError {
     IoError(io::Error),
     FromUtf8Error(string::FromUtf8Error),
-    UnrecognizedElementType(u8),
+    UnrecognizedElementType { key: String, element_type: u8 },
```
Comment (patrickfreed) [inline src/decoder/error.rs:10]: If we ever publicly support decoding `Bson` from a reader without it being part of a `Document`, this key field could present a problem. Though I honestly think the empty string or some other specific value like "TOP LEVEL" wouldn't be that bad. We could always add a `TopLevelUnrecognizedElementType` error as an alternative.

Comment (saghm) [reply src/decoder/error.rs:10]: Could we just make key an `Option`? Alternately, we could name this variant something specific to the fact that it's for values in documents to leave room for us to add a separate variant for unnested values later.

Comment (patrickfreed) [reply src/decoder/error.rs]: I'd rather use separate cases than `Option` personally. I renamed it to `UnrecognizedDocumentElementType`, let me know if you think that works.

Comment (saghm) [reply src/decoder/error.rs]: Yep, works for me

**Pull request merged**